keps/sig-scheduling/5007-device-attach-before-pod-scheduled/README.md

```diff
@@ -0,0 +1,1107 @@
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+**What if the scheduler restarts while the DRA plugin is waiting for the device(s) to be bound?**
+
+The scheduler's restart should not pose an issue, as the decision to wait is based on the Conditions of the ResourceClaim.
+After a scheduler restart, if the device attachment is not yet complete, the scheduler will wait again at PreBind.
+If the attachment is complete, it will pass through PreBind.
+
+**Pods which are not bound yet (in api-server) and not unschedulable (in api-server) are not visible by cluster autoscaler, so there is a risk that the node will be turned down.**
+
+Regarding collaboration with the Cluster Autoscaler, using node nomination can address the issue.
+This issue needs to be resolved before the beta is released.
+
+**The in-flight events cache may grow too large when waiting in PreBind.**
+
+To address the PreBind concern, the solution is to modify the scheduling framework to flush the in-flight events cache before PreBind. 
+This prevents issues in the scheduling queue caused by keeping pods at PreBind for an extended period.
+This issue will be addressed separately as outlined in kubernetes/kubernetes#129967.
+This issue needs to be resolved before the beta is released.
+
+## Design Details
+
+### DRA Scheduler Plugin Design Overview
+
+This document outlines the design of the DRA Scheduler Plugin, focusing on the handling of fabric devices.
+Key additions include new attributes for device identification, enhancements to `AllocatedDeviceStatus`, and the process for handling `ResourceSlices` upon attachment failure.
+The composable controller design is also discussed, emphasizing efficient utilization of fabric devices.
+
+![proposal](proposal.jpg)
+
+#### Device Attribute Additions
+
+To indicate whether a device is a fabric device, an attribute is added to the `Basic` within `Device`.
+This attribute will be used by the controller that exposes the `ResourceSlice` to notify whether the device is a fabric device.
+To avoid impacting existing DRA functionality, the default value of this attribute is set to `false`.
+
+```go
+// Device represents one individual hardware instance that can be selected based
+// on its attributes. Besides the name, exactly one field must be set.
+type Device struct {
+    // Name is unique identifier among all devices managed by
+    // the driver in the pool. It must be a DNS label.
+    //
+    // +required
+    Name string
+
+    // Basic defines one device instance.
+    //
+    // +optional
+    // +oneOf=deviceType
+    Basic *BasicDevice
+}
+
+// BasicDevice represents a basic device instance.
+type BasicDevice struct {
+    // Attributes defines the set of attributes for this device.
+    // The name of each attribute must be unique in that set.
+    //
+    // The maximum number of attributes and capacities combined is 32.
+    //
+    // +optional
+    Attributes map[QualifiedName]DeviceAttribute
```

Comment:
Questions around the API design:
- Any specific reason for this name? For me, it'd be natural if we could call it `Labels` or `Annotations`, which is familiar names within K8s APIs. Is Attribute a more common name for wg/device-management domain?
- Any reason that we cannot do `map[string]string` (which you had before) and rather you prefer to do `map[QualifiedName]DeviceAttribute`? `QualifiedName` is fine. But, `DeviceAttribute` looks weird `boolValue: true` is unfamiliar API, and also maybe we'll have to get more fields to cover other types and stuff in the future, which could bring a troublesomeness here.

Reply:
This is from core DRA. Device attributes are not all just strings, hence the `DeviceAttribute` one-of struct.

Reply:
Okay, if both are coming from the existing concept, then that's fine

Reply:
This field is no longer relevant for the KEP.

You can replace it with `...` to indicate that `BasicDevice` has some other fields besides the ones you add here.